staking/economics/reward.go

```diff
@@ -41,49 +49,73 @@ func adjust(amount numeric.Dec) numeric.Dec {
 func Adjustment(percentageStaked numeric.Dec) (numeric.Dec, numeric.Dec) {
 	howMuchOff := targetStakedPercentage.Sub(percentageStaked)
 	adjustBy := adjust(
-		howMuchOff.MulTruncate(numeric.NewDec(100)).Mul(dynamicAdjust),
+		howMuchOff.MulTruncate(oneHundred).Mul(dynamicAdjust),
 	)
 	return howMuchOff, adjustBy
 }
 
-// WhatPercentStakedNow ..
-func WhatPercentStakedNow(
+// Snapshot returns
+// (soFarDoledOut, stakedNow, stakedPercentage, error)
+func Snapshot(
 	beaconchain engine.ChainReader,
 	timestamp int64,
-) (*big.Int, *numeric.Dec, error) {
-	stakedNow := numeric.ZeroDec()
-	// Only active validators' stake is counted in stake ratio because only their stake is under slashing risk
+	includeAPRs bool,
+) (*big.Int, []computedAPR, *numeric.Dec, error) {
+	stakedNow, rates, junk :=
+		numeric.ZeroDec(), []computedAPR{}, numeric.ZeroDec()
+	// Only active validators' stake is counted in
+	// stake ratio because only their stake is under slashing risk
 	active, err := beaconchain.ReadActiveValidatorList()
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
+	}
+	if includeAPRs {
+		rates = make([]computedAPR, len(active))
 	}
-
 	soFarDoledOut, err := beaconchain.ReadBlockRewardAccumulator(
 		beaconchain.CurrentHeader().Number().Uint64(),
 	)
 
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	dole := numeric.NewDecFromBigInt(soFarDoledOut)
 
 	for i := range active {
 		wrapper, err := beaconchain.ReadValidatorInformation(active[i])
 		if err != nil {
-			return nil, nil, err
+			return nil, nil, nil, err
+		}
+		total := wrapper.TotalDelegation()
+		stakedNow = stakedNow.Add(numeric.NewDecFromBigInt(total))
+		if includeAPRs {
+			rates[i] = computedAPR{active[i], total, junk, numeric.ZeroDec()}
 		}
-		stakedNow = stakedNow.Add(
-			numeric.NewDecFromBigInt(wrapper.TotalDelegation()),
-		)
 	}
-	percentage := stakedNow.Quo(totalTokens.Mul(
+
+	circulatingSupply := totalTokens.Mul(
 		reward.PercentageForTimeStamp(timestamp),
-	).Add(dole))
+	).Add(dole)
+
+	for i := range rates {
+		rates[i].StakeRatio = numeric.NewDecFromBigInt(
```

Comment:
This is wrong. The APR in the formula I wrote is for the general APR for staking assuming all stake are fully utilized. It's not for individual validator's yield. Individual validator's yield should be dependent on their uptime (basically their actual block reward earned). Basically, individual validator's yield <= general APR